File: spec/lib/Binary_Search_Spec.rb
require 'spec_helper'

describe 'Binary_Search' do

  before :each do
    @bs = Binary_Search.new
    @collection = My_Container.new 100
  end
  describe 'search' do
    it 'should return the index of the value specified' do
      @bs.search(@collection, 5).should eq(5)
      @bs.search(@collection, 0).should eq(0)
      @bs.search(@collection, 99).should eq(99)
      @bs.search(@collection, 63).should eq(63)
    end

    it 'should return nil if the value is not found' do
      @bs.search(@collection, 100).should eq(nil)
      @bs.search(@collection, 135341).should eq(nil)
      @bs.search(@collection, -1).should eq(nil)
    end
  end

  describe 'recursive_search' do
    it 'should return the index of the value specified' do
      @bs.recursive_search(@collection, 5).should eq(5)
      @bs.recursive_search(@collection, 0).should eq(0)
      @bs.recursive_search(@collection, 99).should eq(99)
      @bs.recursive_search(@collection, 63).should eq(63)
    end

    it 'should return nil if the value is not found' do
      @bs.recursive_search(@collection, 100).should eq(nil)
      @bs.recursive_search(@collection, 135341).should eq(nil)
      @bs.recursive_search(@collection, -1).should eq(nil)
    end
  end
end

File: app/spec/javascripts/search/BinarySearchSpec.js
describe('BinarySearch', function() {
  var bs, searchList;

  beforeEach(function() {
    searchList = [0,1,2,3,4,5,6,7,8,9];
    setFixtures($("<div class='demo'></div>"));
    bs = new BinarySearch($('.demo').get(0));
  });

  describe('init', function() {
    it('should insert the elements into the dom', function() {
      expect($('.collection > li')).toHaveLength(searchList.length);

      forEach($('.collection > li'), function(li, index) {
        expect(parseInt(li.innerHTML)).toEqual(index);
      });
    });
  });

  describe('search', function() {
    it('should return an instruction set that finds the element', function() {
      var queue = bs.search(8);

      expect(queue.length).toBeGreaterThan(0);
      execute(queue);

      var queue = bs.search(8);
      execute(queue);

      expect(parseInt($('.index8').html())).toEqual(8);
      expect($('.index8')).toHaveClass('found');
    });
  });

  describe('cleanElements', function() {
    it('should reset the elements', function() {
      var queue = bs.search(5);
      execute(queue);

      bs.cleanElements();
      forEach($('.collection > li'), function(li) {
        expect(li).not.toHaveClass('found');
        expect(li).not.toHaveClass('search');
        expect(li).not.toHaveClass('ignore');
      });
    });
  });
});
